---
title: "Microbiome - Genomics Final Project"
author: "Peter D. Gandy"
date: "April 18th, 2023"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Biological Science 469 Genomic Data Analysis Spring Final Project

This project will use QIIME 2 to perform an analysis of human microbiome samples from two individuals at four body sites (right and left palm, gut, and tongue) at five time points, the first of which immediately followed antibiotic usage. A study based on these samples was originally published in Caporaso et al. (2011). The data used in this project were sequenced on an Illumina HiSeq using the Earth Microbiome Project hypervariable region 4 (V4) 16S rRNA sequencing protocol.

Reference:
Caporaso JG, Lauber CL, Costello EK, Berg-Lyons D, Gonzalez A, Stombaugh J, Knights D, Gajer P, Ravel J, Fierer N, Gordon JI, Knight R. Moving pictures of the human microbiome. Genome Biol. 2011;12(5):R50. doi: 10.1186/gb-2011-12-5-r50. PMID: 21624126; PMCID: PMC3271711.
```{r}
# Loading Packages#
p1 <- c("tidyverse", "vegan", "BiocManager")
p2 <- c("phyloseq", "ANCOMBC", "DESeq2", "ComplexHeatmap")
load_package <- function(p) {
  if (!requireNamespace(p, quietly = TRUE)) {
    ifelse(p %in% p1, 
           install.packages(p, repos = "http://cran.us.r-project.org/"), 
           BiocManager::install(p))
  }
  library(p, character.only = TRUE, quietly = TRUE)
}
invisible(lapply(c(p1,p2), load_package))

# Data management
# vignette("phyloseq-basics")
# vignette("phyloseq-analysis")
library(ape)
library(maps)
library(phytools)
library(phyloseq)
library(SummarizedExperiment)
library(plyr)
library(dplyr)
# Graphics and tables#
library(ggplot2)
library(cowplot)
library(kableExtra)
library(vegan)
library(reprex)
library(readxl)
library(phyloseq)
library(tidytree)
library(kableExtra)

# Preparation#
# Download Files#
#rooted-tree.qza
#sample-metadata.tsv
#table.qza
#taxonomy.qza

# conda activate qiime2-2020.11
# for i in *.qza; do
# qiime tools export --input-path $i --output-path .
# done
# biom convert -i feature-table.biom -o feature-table.tsv --to-tsv
# conda deactivate

# feature-table.tsv
# sample-metadata.tsv
# taxonomy.tsv
# tree.nwk}
```

```{r}
# BEFORE PRUNING
#  Start by loading data in R and converting it to the proper format (matrix/data.frame)#
# OTU DATA STRUCTURE (I)#
otu <- read.table(file = "C://Users//pgand//OneDrive//Documents//BioSci 469 Genomic Data Analysis//Project//feature-table.tsv", sep = "\t", header = T, row.names = 1, skip = 1, comment.char = "")
otu[1:2, 1:6]
```
```{r}
# TAXONOMY DATA STRUCTURE (I)#
taxonomy <- read.table(file = "C://Users//pgand//OneDrive//Documents//BioSci 469 Genomic Data Analysis//Project//taxonomy.tsv", sep = "\t", header = T ,row.names = 1)
taxonomy[1:2, ]
```
```{r}
# METADATA DATA STRUCTURE (I)#
metadata <- read.table(file = ("C://Users//pgand//OneDrive//Documents//BioSci 469 Genomic Data Analysis//Project//metadata.tsv"), sep = "\t", header = T, row.names = 1)
metadata[1:2, ]
```
```{r}
# PHYLOGENTIC DATA STRUCTURE (I)#
tree = ape::read.tree("C://Users//pgand//OneDrive//Documents//BioSci 469 Genomic Data Analysis//Project//tree.nwk")
phy_tree(tree)
```
```{r}
# Clean the taxonomy, Green genes format before cleaning#
tax <- taxonomy %>%
  select(Taxon) %>% 
  separate(Taxon, c("Kingdom", "Phylum", "Class", "Order", "Family", "Genus", "Species"), "; ")
dim(tax)
head(tax)
```
```{r}
# Green Genes format after cleaning
tax.clean <- data.frame(row.names = row.names(tax),
                        Kingdom = str_replace(tax[,1], "k__",""),
                        Phylum = str_replace(tax[,2], "p__",""),
                        Class = str_replace(tax[,3], "c__",""),
                        Order = str_replace(tax[,4], "o__",""),
                        Family = str_replace(tax[,5], "f__",""),
                        Genus = str_replace(tax[,6], "g__",""),
                        Species = str_replace(tax[,7], "s__",""),
                        stringsAsFactors = FALSE)

tax.clean[is.na(tax.clean)] <- ""
tax.clean[tax.clean=="__"] <- ""
dim(tax.clean)
head(tax.clean)

for (i in 1:nrow(tax.clean)){
  if (tax.clean[i,7] != ""){
    tax.clean$Species[i] <- paste(tax.clean$Genus[i], tax.clean$Species[i], sep = " ")
  } else if (tax.clean[i,2] == ""){
    kingdom <- paste("Unclassified", tax.clean[i,1], sep = " ")
    tax.clean[i, 2:7] <- kingdom
  } else if (tax.clean[i,3] == ""){
    phylum <- paste("Unclassified", tax.clean[i,2], sep = " ")
    tax.clean[i, 3:7] <- phylum
  } else if (tax.clean[i,4] == ""){
    class <- paste("Unclassified", tax.clean[i,3], sep = " ")
    tax.clean[i, 4:7] <- class
  } else if (tax.clean[i,5] == ""){
    order <- paste("Unclassified", tax.clean[i,4], sep = " ")
    tax.clean[i, 5:7] <- order
  } else if (tax.clean[i,6] == ""){
    family <- paste("Unclassified", tax.clean[i,5], sep = " ")
    tax.clean[i, 6:7] <- family
  } else if (tax.clean[i,7] == ""){
    tax.clean$Species[i] <- paste("Unclassified ",tax.clean$Genus[i], sep = " ")
  }
}
```

```{r}
# BUILD PHYLOSEQ OBJECT
# OTU TABLE: matrix-like object
OTU = otu_table(as.matrix(otu), taxa_are_rows = TRUE)
head(OTU)
```
```{r}
# TAXONOMY TABLE: matrix-like object#
TAX = tax_table(as.matrix(tax.clean))
head(tax_table(TAX))
```
```{r}
# SAMPLE DATA: Data.frame-like object#
SAMPLE <- sample_data(metadata)
head(sample_data(SAMPLE), n = 4)
```
```{r}
# PHY TREE: phylo-class (tree) object#
phy_tree(tree)
```

```{r}
# Building Phyloseq Object#
ps <- merge_phyloseq(otu_table(OTU, taxa_are_rows = TRUE),
                 tax_table(TAX),
                 phy_tree(tree),
                 sample_data(SAMPLE))
ps
```

```{r}
# There are 131 different families#
get_taxa_unique(ps, "Family")
```

```{r}
# ACCESSORS#
ntaxa(ps)
nsamples(ps)
sample_names(ps)[1:5]
rank_names(ps)
sample_variables(ps)
otu_table(ps)[1:5, 1:5]
tax_table(ps)[1:5, 1:4]
myTaxa = names(sort(taxa_sums(ps), decreasing = TRUE)[1:10])
ex1 = prune_taxa(myTaxa, ps)


coarse.ps <- tax_glom(ps, "Phylum")
ntaxa(coarse.ps) ## number of different phyla
```

```{r}
# AFTER PRUNING#
# PRUNING UNWANTED TAX#
set.seed(1) # keep result reproductive
ps.rarefied = rarefy_even_depth(ps, rngseed=33612, sample.size=1103, replace=F)
ps.rarefied
```

```{r}
# Relative abundance plots at family level#
# There are 115 different families#
get_taxa_unique(ps.rarefied, "Family")
```

```{r}
# Statistics for the number of observed tags#
alpha.diversity = estimate_richness(ps.rarefied, measures = c("Observed", "Chao1", "Shannon", "InvSimpson"))
head(alpha.diversity)
```

```{r}
data <- cbind(sample_data(ps.rarefied), alpha.diversity)
head(data)
```

```{r}
# Statistics for the number of Observed tags#
wilcox.observed <- pairwise.wilcox.test(alpha.diversity$Observed, 
                                        sample_data(ps.rarefied)$body.site, 
                                        p.adjust.method = "BH")
tab.observed <- wilcox.observed$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()
tab.observed
```

```{r}
# Statistics for the number of Shannon tags#
wilcox.shannon <- pairwise.wilcox.test(alpha.diversity$Shannon, 
                                       sample_data(ps.rarefied)$body.site, 
                                       p.adjust.method = "BH")
tab.shannon <- wilcox.shannon$p.value %>%
  as.data.frame() %>%
  tibble::rownames_to_column(var = "group1") %>%
  gather(key="group2", value="p.adj", -group1) %>%
  na.omit()
tab.shannon
```

```{r}
# BETA DIVERSITY - assesses the dissimilarity between ecosystems#
dist = phyloseq::distance(ps.rarefied, method="bray")
ordination = ordinate(ps.rarefied, method="PCoA", distance=dist)
```

```{r}
# PCoA - MDS + Bray−Curtis#
ord <- ordinate(ps.rarefied, method = "MDS", distance = "bray")
```

```{r}
# PCoA - MDS + Jaccard#
ord <- ordinate(ps.rarefied, method = "MDS", distance = "jaccard")
```

```{r}
#PCoA - MDS + unifrac#
ord <- ordinate(ps.rarefied, method = "MDS", distance = "chao")
```

```{r}
#PCoA - MDS + wunifrac#
ord <- ordinate(ps.rarefied, method = "MDS", distance = "wunifrac")
```

```{r}
#Ordination#
dist.bc <- phyloseq::distance(ps.rarefied, "bray")
kin.mds <- metaMDS(dist.bc, trace = 0)
```

```{r}
#Vector fitting#
ef <- envfit(kin.mds, sample_data(ps.rarefied))
```

```{r}
#PERMANOVA/ADONIS2 - euclidean#
metadata <- data.frame(sample_data(ps.rarefied))
test.adonis <- adonis2(dist ~ body.site, data = metadata, sim.function='vegdist',sim.method='euclidian',p.adjust.m='holm')
test.adonis
```

```{r}
#PERMANOVA/ADONIS2 - Manhattan#
metadata <- data.frame(sample_data(ps.rarefied))
test.adonis <- adonis2(dist ~ body.site, data = metadata, sim.function='daisy',sim.method='manhattan',p.adjust.m='bonferroni')
test.adonis
```

```{r}
#NMDS and ANOSIM#
#PCoA - NMDS + Chao#
dist = phyloseq::distance(ps.rarefied, method="chao", binary = TRUE)
ordination = ordinate(ps.rarefied, method="NMDS", distance=dist)
```

```{r}
#PCoA - NMDS + Bray#
dist = phyloseq::distance(ps.rarefied, method="bray", binary = TRUE)
ordination = ordinate(ps.rarefied, method="NMDS", distance=dist)
```

```{r}
#PCoA - NMDS + Jaccard#
dist = phyloseq::distance(ps.rarefied, method="jaccard", binary = TRUE)
ordination = ordinate(ps.rarefied, method="NMDS", distance=dist)
```

```{r}
# ANOSIM#
metadata <- data.frame(sample_data(ps.rarefied))
anosim(dist, metadata$body.site)
```

```{r}
#PAIRWISE ANOSIM#
cbn <- combn(x=unique(metadata$body.site), m = 2)
p <- c()

for(i in 1:ncol(cbn)){
  ps.subs <- subset_samples(ps.rarefied, body.site %in% cbn[,i])
  metadata_sub <- data.frame(sample_data(ps.subs))
  permanova_pairwise <- anosim(phyloseq::distance(ps.subs, method="jaccard", binary = TRUE), 
                               metadata_sub$body.site)
  p <- c(p, permanova_pairwise$signif[1])
}

p.adj <- p.adjust(p, method = "BH")
p.table <- cbind.data.frame(t(cbn), p=p, p.adj=p.adj)
p.table
```

```{r}
# Abundance Phylum#
ps.rel = transform_sample_counts(ps, function(x) x/sum(x)*100)
```

```{r}
# Agglomerate Taxa#
glom <- tax_glom(ps.rel, taxrank = 'Phylum', NArm = FALSE)
ps.melt <- psmelt(glom)
```

```{r}
# Change to character for easy-adjusted level#
ps.melt$Phylum <- as.character(ps.melt$Phylum)
ps.melt <- ps.melt %>%
  group_by(body.site, Phylum) %>%
  mutate(median=median(Abundance))
```

```{r}
#Select group median > 1#
keep <- unique(ps.melt$Phylum[ps.melt$median > 1])
ps.melt$Phylum[!(ps.melt$Phylum %in% keep)] <- "< 1%"
```

```{r}
#To get the same rows together#
ps.melt_sum <- ps.melt %>%
  group_by(Sample,body.site,Phylum) %>%
  summarise(Abundance=sum(Abundance))
```
```{r fig.asp = 0.8, fig.width = 10}
# Abundance Phylum
p <- ggplot(ps.melt_sum, aes(x = Sample, y = Abundance, fill = Phylum)) + 
  geom_bar(stat = "identity", aes(fill=Phylum)) + 
  labs(x="", y="%") +
  facet_wrap(~body.site, scales= "free_x", nrow=1) +
  theme_classic() + 
  theme(strip.background = element_blank(), 
        axis.text.x.bottom = element_text(angle = -90))
p
```

```{r}
#Abundance Genus
ps.rel = transform_sample_counts(ps, function(x) x/sum(x)*100)
```

```{r}
#Agglomerate taxa#
glom <- tax_glom(ps.rel, taxrank = 'Genus', NArm = FALSE)
ps.melt <- psmelt(glom)
```

```{r}
#Change to character for easy-adjusted level#
ps.melt$Genus <- as.character(ps.melt$Genus)
ps.melt <- ps.melt %>%
  group_by(body.site, Genus) %>%
  mutate(median=median(Abundance))
```

```{r}
#Select group mean > 2.5#
keep <- unique(ps.melt$Genus[ps.melt$median > 2.5])
ps.melt$Genus[!(ps.melt$Genus %in% keep)] <- "< 2.5%"
```

```{r}
#To get the same rows together
ps.melt_sum <- ps.melt %>%
  group_by(Sample,body.site,Genus) %>%
  summarise(Abundance=sum(Abundance))
```

```{r}
#Differential Abundance Analysis#
sample_data(ps)$body.site <- as.factor(sample_data(ps)$body.site) # factorize for DESeq2
ps.taxa <- tax_glom(ps, taxrank = 'Species', NArm = FALSE)
```

```{r}
#Pairwise comparison between gut and tongue#
ps.taxa.sub <- subset_samples(ps.taxa, body.site %in% c("gut", "tongue"))
```

```{r}
#Filter sparse features, with > 90% zeros#
ps.taxa.pse.sub <- prune_taxa(rowSums(otu_table(ps.taxa.sub) == 0) < ncol(otu_table(ps.taxa.sub)) * 0.9, ps.taxa.sub)
ps_ds = phyloseq_to_deseq2(ps.taxa.pse.sub, ~ body.site)
```

```{r}
#Use alternative estimator on a condition of "every gene contains a sample with a zero"
ds <- estimateSizeFactors(ps_ds, type="poscounts")
ds = DESeq(ds, test="Wald", fitType="parametric")
alpha = 0.05 
res = results(ds, alpha=alpha)
res = res[order(res$padj, na.last=NA), ]
taxa_sig = rownames(res[1:20, ]) #select bottom 20 with lowest p.adj values#
ps.taxa.rel <- transform_sample_counts(ps, function(x) x/sum(x)*100)
ps.taxa.rel.sig <- prune_taxa(taxa_sig, ps.taxa.rel)
```

```{r}
#Only keep gut and tongue samples#
ps.taxa.rel.sig <- prune_samples(colnames(otu_table(ps.taxa.pse.sub)), ps.taxa.rel.sig)
ps.taxa.rel.sig
```

```{r}
#HEATMAPS#
matrix <- as.matrix(data.frame(otu_table(ps.taxa.rel.sig)))
rownames(matrix) <- as.character(tax_table(ps.taxa.rel.sig)[, "Species"])
metadata_sub <- data.frame(sample_data(ps.taxa.rel.sig))
```

```{r}
#Define the annotation color for columns and rows#
annotation_col = data.frame(
  Subject = as.factor(metadata_sub$subject), 
  `Body site` = as.factor(metadata_sub$body.site), 
  check.names = FALSE
)
rownames(annotation_col) = rownames(metadata_sub)

annotation_row = data.frame(
  Phylum = as.factor(tax_table(ps.taxa.rel.sig)[, "Phylum"])
)
rownames(annotation_row) = rownames(matrix)
```

```{r}
#ann_color should be named vectors
phylum_col = RColorBrewer::brewer.pal(length(levels(annotation_row$Phylum)), "Paired")
names(phylum_col) = levels(annotation_row$Phylum)
ann_colors = list(
  Subject = c(`subject-1` = "red", `subject-2` = "blue"),
  `Body site` = c(gut = "purple", tongue = "yellow"),
  Phylum = phylum_col
)
```



## Including Plots

You can also embed plots, for example:
  
```{r Phy_tree, echo=FALSE}
plot(phy_tree(ex1), show.node.label = TRUE)
```

```{r Alpha Diversity plot_richness, echo=FALSE}
# Alpha Diversity - How diverse a sequenced community is#
plot_richness(ps.rarefied, x="body.site", measures=c("Observed", "Chao1", "Shannon", "InvSimpson"))+
  geom_boxplot() +
  theme_classic() +
  theme(strip.background = element_blank(), axis.text.x.bottom = element_text(angle = -90))
```

```{r plot MDS + Bray−Curtis, echo=FALSE}
# PCoA PLOT MDS + Bray−Curtis#
p <- plot_ordination(ps.rarefied, ord, color = "body.site", shape = "Bande")
p <- p + theme_bw() + ggtitle("MDS + BC") ## add title and plain background
p <- p + stat_ellipse(aes(group = body.site)) ## add ellipses around each time level
plot(p)
```

```{r plot MDS + Jaccard, echo=FALSE}
# PCoA PLOT MDS + Jaccard#
p <- plot_ordination(ps.rarefied, ord, color = "body.site", shape = "Bande")
p <- p + theme_bw() + ggtitle("MDS + Jaccard") ## add title and plain background
p <- p + stat_ellipse(aes(group = body.site)) ## add ellipses around each time level
plot(p)
```

```{r plot MDS + Unifrac, echo=FALSE}
# PCoA PLOT MDS + Chao #
p <- plot_ordination(ps.rarefied, ord, color = "body.site", shape = "Bande")
p <- p + theme_bw() + ggtitle("MDS + Chao") ## add title and plain background
p <- p + stat_ellipse(aes(group = body.site)) ## add ellipses around each time level
plot(p)
```

```{r plot MDS + Wunifrac, echo=FALSE}
# PCoA PLOT MDS + Wunifrac#
p <- plot_ordination(ps.rarefied, ord, color = "body.site", shape = "Bande")
p <- p + theme_bw() + ggtitle("MDS + wUF") ## add title and plain background
p <- p + stat_ellipse(aes(group = body.site)) ## add ellipses around each time level
plot(p)
```

```{r plot significant variables, echo=FALSE}
# Plot only most significant variables#
plot(kin.mds)
plot(ef, p.max = 0.001)
```

```{r plot NMDS + Chao, echo=FALSE}
# # PCoA PLOT NMDS + Chao#
p <- plot_ordination(ps.rarefied, ordination, color = "body.site", shape = "Bande")
p <- p + theme_bw() + ggtitle("NMDS + Chao") ##add title and plain background
p <- p + stat_ellipse(aes(group = body.site)) ## add ellipses around each time level
plot(p)
```

```{r plot NMDS + BC, echo=FALSE}
# PCoA PLOT NMDS + BC#
p <- plot_ordination(ps.rarefied, ordination, color = "body.site", shape = "Bande")
p <- p + theme_bw() + ggtitle("NMDS + BC") ## add title and plain background
p <- p + stat_ellipse(aes(group = body.site)) ## add ellipses around each time level
plot(p)
```

```{r plot NMDS + Jaccard, echo=FALSE}
# PCoA PLOT NMDS + Jaccard#
p <- plot_ordination(ps.rarefied, ordination, color = "body.site", shape = "Bande")
p <- p + theme_bw() + ggtitle("NMDS + Jaccard") ## add title and plain background
p <- p + stat_ellipse(aes(group = body.site)) ## add ellipses around each time level
plot(p)
```

```{r fig.asp = 0.8, fig.width = 10}
# Abundance Genus
ggplot(ps.melt_sum, aes(x = Sample, y = Abundance, fill = Genus)) + 
  geom_bar(stat = "identity", aes(fill=Genus)) + 
  labs(x="", y="%") +
  facet_wrap(~body.site, scales= "free_x", nrow=1) +
  theme_classic() + 
  theme(legend.position = "right", 
        strip.background = element_blank(), 
        axis.text.x.bottom = element_text(angle = -90))
```

```{r fig.asp = 0.8, fig.width = 10}
#HEATMAP
ComplexHeatmap::pheatmap(matrix, scale= "row", 
                         annotation_col = annotation_col, 
                         annotation_row = annotation_row, 
                         annotation_colors = ann_colors)
```

```{r}
sessionInfo()
```

Note that the `echo = FALSE` parameter was added to the code chunk to prevent printing of the R code that generated the plot.
